### DISCUSSION TITLE
client.run_javascript() in separate thread gives "JavaScript did not respond in time"

### DISCUSSION THREAD

**Discussion by ItsCubeTime** [zauberzeug/nicegui#920] [Q&A] [answered]
### Question

Hi 👋 Continuing the discussion from https://github.com/zauberzeug/nicegui/discussions/823 Im struggling to run any JS in my own thread:
```py
from nicegui import ui
import nicegui
import typing
import threading
import queue
import asyncio
def runSync(coroutine: typing.Coroutine, dontWait = False):
    """Takes a coroutine and runs it syncronously, giving you the returnvalue.

    Current implementation spins up a new thread, creates a new asyncio event loop in there and executes the corotine
    """

    async def threadAsyncCb():
        return await coroutine
    def threadMain(out_queue1):
        loop = asyncio.new_event_loop()  # Because were in a new thread
        # we can be confident theres no already running event loop.
        out_queue1.put(loop.run_until_complete(threadAsyncCb()))
        loop.close()

    out_queue1 = queue.Queue()
    t1 = threading.Thread(target=threadMain, args=(out_queue1,))
    t1.start()
    if dontWait:
        return t1
    t1.join()
    return out_queue1.get()
client = nicegui.globals.get_client()
ui.button(
    "Run JavaScript!", 
    on_click=lambda: runSync(client.run_javascript("alert(`Nicegui is 💝`)"))
          )
ui.run(native=True)
```

```
python -u "c:\Users\olliv\Desktop\Art And Development\toga\toga dev test env\niceguiRunJavascriptInThread.py"
NiceGUI ready to go on http://127.0.0.1:8000
Exception in thread Thread-1 (threadMain):
Traceback (most recent call last):
  File "C:\Users\olliv\AppData\Local\Programs\Python\Python310\lib\threading.py", line 1009, in _bootstrap_inner
    self.run()
  File "C:\Users\olliv\AppData\Local\Programs\Python\Python310\lib\threading.py", line 946, in run
    self._target(*self._args, **self._kwargs)
  File "c:\Users\olliv\Desktop\Art And Development\toga\toga dev test env\niceguiRunJavascriptInThread.py", line 19, in threadMain
    out_queue1.put(loop.run_until_complete(threadAsyncCb()))
  File "C:\Users\olliv\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 641, in run_until_complete
    return future.result()
  File "c:\Users\olliv\Desktop\Art And Development\toga\toga dev test env\niceguiRunJavascriptInThread.py", line 14, in threadAsyncCb
    return await coroutine
  File "C:\Users\olliv\AppData\Local\Programs\Python\Python310\lib\site-packages\nicegui\client.py", line 128, in run_javascript
    raise TimeoutError('JavaScript did not respond in time')
TimeoutError: JavaScript did not respond in time
```

Do you guys know whats going on here? 

Im testing on Windows 11

**Comment by rodja** [comment, marked as answer]
All UI calls (including `run_javascript`) must be executed on the main thread (and the event loop which is running there). Why do you want to "take a coroutine and run it synchronously"? A simplified version of your code above could look like this

```py
async def js():
    await ui.run_javascript('alert(`NiceGUI is 💝`)', respond=False)

ui.button('Run JavaScript!', on_click=js)
```

But maybe your real world scenario is a bit different...

**Reply by ItsCubeTime** [reply]
Thanks rodja, when you say main thread you mean the thread created for the particular connected client, right? Or does all outbound traffic (to all the connected clients) travel through the same thread?

As for the real world example:
![image](https://github.com/zauberzeug/nicegui/assets/20190653/7e29bfca-8de4-4516-823b-fa9ed89714e2)
I dont want my "getPlatform" method to be an async method as I want it to be callable anywhere, in any context. Making it async would limit the amount of places where I can execute it.

**Reply by rodja** [reply]
NiceGUI (and FastAPI, and underlying Starlette + python-socketio) are Async frameworks. That means they solve concurrency without threading. Instead there is an event loop which consumes pieces of computation one at a time with the assumption that each piece is very fast (ideally < 10 ms). See https://fastapi.tiangolo.com/async/ for more details. In short: yes, all traffic is handled by a single thread.
When you call JavaScript, there is quite some communication happening in the background. Without async/await this communication would block the main thread, stopping all other executions. If no value needs to be returned you could use `nicegui.background_task.create`. But in your `getPlatform` example this is not an option.

>  I want it to be callable anywhere, in any context

That feels like not the right thing to do. Even if the blocking problem described above would not be an issue, you already figured out that you need a `Client` to call JavaScript. `get_client` tries to find the correct `client` by looking into the NiceGUI context... which could be anything when called from another thread.

Maybe a better solution would be to determine the platform once when the connection is established and store the result somewhere so you can later access it from different threads?
